### PR TITLE
Diagnosis manager implementation in the master

### PR DESCRIPTION
**Pull request by samplise**
### What changes were proposed in this pull request?

Implement the diagnosis manager in the master.

### Why are the changes needed?

The diagnosis manager is responsible for collecting diagnosis data from workers. Then it identifies and handles training faults.

### Does this PR introduce any user-facing change?

No.

### How was this patch tested?

Unit test.

### CODE REVIEW

File: dlrover/python/master/diagnosis/inferencechain/common.py [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+        return True
+    else:
+        return False
+
+
+def combine_inferences(
+    inferences1: List[Inference], inferences2: List[Inference]
+) -> List[Inference]:
+    inferences = []
+    for inference2 in inferences2:
```
Comment (BalaBalaYi) [inline dlrover/python/master/diagnosis/inferencechain/common.py:74]:
```
inferences = list(set(inferences1 + inferences2))
```

File: dlrover/python/master/diagnosis/inferencechain/inference_chain.py [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+    Inference,
+    InferenceOperator,
+    combine_inferences,
+)
+from dlrover.python.master.diagnosis.operator.operator import (
+    register_operators,
+)
+
+
+class InferenceChain:
```
Comment (BalaBalaYi) [inline dlrover/python/master/diagnosis/inferencechain/inference_chain.py:28]: Need annotations.

File: dlrover/python/master/diagnosis/inferencechain/inference_chain.py [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+                    operator = self.get_operator(inference)
+                    infs = operator.infer(inferences)
+                    if len(infs) > 0:
+                        has_new_inference = True
+                        new_infs = combine_inferences(new_infs, infs)
+                    else:
+                        new_infs.append(inference)
+                except Exception as e:
+                    logger.exception(e)
+                    new_infs.append(inference)
```
Comment (BalaBalaYi) [inline dlrover/python/master/diagnosis/inferencechain/inference_chain.py:50]: Why still append inference in 'catch'?

File: dlrover/python/master/diagnosis/inferencechain/inference_chain.py [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+
+            if has_new_inference:
+                inferences = new_infs
+            else:
```
Comment (BalaBalaYi) [inline dlrover/python/master/diagnosis/inferencechain/inference_chain.py:54]: Is there any possible 'has_new_inference' always false?

**Review by BalaBalaYi — CHANGES_REQUESTED** [review]

File: dlrover/python/common/grpc.py [hunk 5]
```diff
@@ -455,14 +455,14 @@ class NodeCheckpointState(Message):
 
 @dataclass
 class DiagnosisTrainingLog(Message):
-    timestamp: int = 0
```
[the remark this answers is not in the document]
Comment (samplise) [reply dlrover/python/common/grpc.py:458]: Here I use `time.now().timestamp()` to get the timestamp that returns a float.

File: dlrover/python/common/diagnosis.py [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+    def __init__(self):
+        pass
+
+    @abstractmethod
+    def get_timestamp(self) -> float:
+        pass
+
+
+class CudaLog(DiagnosisData):
+    def __init__(self, timestamp: float):
```
[the remark this answers is not in the document]
Comment (samplise) [reply dlrover/python/common/diagnosis.py:34]: Currently those class definitions are not finalized. It is possible to modify the definition in the future.

File: dlrover/python/common/diagnosis.py [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+# limitations under the License.
+
+from abc import ABCMeta, abstractmethod
+from datetime import datetime
+
+
+class DiagnosisDataType:
+    CUDALOG = "cuda_log"
+    TRAININGLOG = "training_log"
+    CHIPMETRICES = "chip_metrics"
```
[the remark this answers is not in the document]
Comment (samplise) [reply dlrover/python/common/diagnosis.py:21]: Here chip metrics include temperature, utilization, etc. I think metrics are more accurate.

File: dlrover/python/elastic_agent/datacollector/cuda_log_collector.py [hunk 8]
```diff
@@ -11,24 +11,18 @@
...
-        self.timestamp = int(round(datetime.now().timestamp()))
-
-
 class CudaLogCollector(DataCollector):
     def __init__(self, *args, **kwargs):
         pass
 
     def collect_data(self) -> object:
-        log = CudaLog()
+        log = CudaLog(0)
```
[the remark this answers is not in the document]
Comment (samplise) [reply dlrover/python/elastic_agent/datacollector/cuda_log_collector.py:25]: It indicates the timestamp is assigned and should use current timestamp.

File: dlrover/python/common/grpc.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (BalaBalaYi) [reply dlrover/python/common/grpc.py:458]: ```time.now()``` -> ```datatime.now()``` ?

File: dlrover/python/master/diagnosis/analyst.py [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+    def __init__(self, data_mgr: DataManager):
+        self.data_manager = data_mgr
+        self.problems = register_problems()
+
+    def observe_training(self) -> List[Inference]:
+        ic = InferenceChain(self.data_manager, self.problems)
+        return ic.infer()
+
+
+def register_problems() -> List[Inference]:
```
[the remark this answers is not in the document]
Comment (samplise) [reply dlrover/python/master/diagnosis/analyst.py:38]: Yes. The diagnosis needs to check if there is any listed problem during training. This function is used to register those problems.

File: dlrover/python/master/diagnosis/diagnosis.py [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+            )
+            thread.start()
+            if thread.is_alive():
+                logger.info("Diagnosis Manager is started")
+        except Exception as e:
+            logger.error(
+                f"Failed to start the diagnosis manager thread. Error: {e}"
+            )
+
+    def stop(self):
```
[the remark this answers is not in the document]
Comment (samplise) [reply dlrover/python/master/diagnosis/diagnosis.py:52]: It is not a final implementation. The current version only reflects the design.

File: dlrover/python/master/diagnosis/diagnosis.py [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+        pass
+
+    def _diagnose_faults(self):
+        logger.info("Start to diagnose faults")
+        while True:
+            observed_problems = self.analyst.observe_training()
+            for problem in observed_problems:
+                logger.info(problem.to_string())
+            time.sleep(180)
```
[the remark this answers is not in the document]
Comment (samplise) [reply dlrover/python/master/diagnosis/diagnosis.py:61]: Sure. I plan to put it in the next pr.

File: dlrover/python/master/diagnosis/inferencechain/common.py [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+
+class InferenceAttribute:
+    ISORNOT = "is_or_not"
+
+
+class InferenceDescription:
+    HANG = "hang"
+
+
+class Inference(object):
```
[the remark this answers is not in the document]
Comment (workingloong) [reply dlrover/python/master/diagnosis/inferencechain/common.py:31]: Use dataclass to wrap the `Inference`.

```Python
from dataclasses import dataclass

@dataclass
class Inference:
    name: str = ""
    attribute: str = ""
    description: str = ""
```

File: dlrover/python/master/diagnosis/inferencechain/inference_chain.py [hunk 13]
```diff
@@ -39,7 +43,9 @@ def infer(self) -> List[Inference]:
             for inference in inferences:
                 try:
                     operator = self.get_operator(inference)
+                    print(operator)
```
Comment (workingloong) [inline dlrover/python/master/diagnosis/inferencechain/inference_chain.py:46]: Remove the line 46 and 48.

**Review by workingloong — APPROVED** [review]
LGTM

**Review by BalaBalaYi — APPROVED** [review]

**Pull request merged**